dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp

```diff
@@ -1163,17 +1164,35 @@ SourceOps DeltaMergeStore::readSourceOps(
         final_num_stream);
 
     SourceOps res;
-    RUNTIME_CHECK(enable_read_thread); // TODO: support keep order
     for (size_t i = 0; i < final_num_stream; ++i)
     {
-        res.push_back(
-            std::make_unique<UnorderedSourceOp>(
-                exec_status_,
+        if (enable_read_thread)
+        {
+            res.push_back(
+                std::make_unique<UnorderedSourceOp>(
+                    exec_status,
+                    read_task_pool,
+                    columns_to_read,
+                    extra_table_id_index,
+                    physical_table_id,
+                    log_tracing_id));
+        }
+        else
+        {
+            res.push_back(std::make_unique<DMSegmentThreadSourceOp>(
```

Comment:
for non-read-pool, the final_num_stream should be `std::max(1, std::min(num_streams, tasks.size()))`.